File: sdk/python/figuard/_embedded/backend.py
"""Embedded backend — serves the FiGuard REST contract in-process against local SQLite.

It implements the same ``(method, path, json, headers) -> server-shaped JSON`` contract the
HTTP transport provides, so ``FiGuardClient``'s methods work UNCHANGED on either backend. The
engine returns decimal-string / snake_case results; this layer reshapes them into the server's
JSON (camelCase, numbers) that the SDK's ``_parse_*`` functions already consume.

Fleet endpoints (delegation, subscriptions, entitlements, webhooks, replay, …) raise
FiGuardCapabilityError — the embedded backend draws the embedded/server line in software.
"""

from __future__ import annotations

import re
import uuid
from typing import Optional

from ..exceptions import FiGuardApiError
from .engine import LiteEngine
from .errors import FiGuardCapabilityError, InvalidParentError

_CONFIRM = re.compile(r"^/api/v1/events/([^/]+)/confirm$")
_FAIL = re.compile(r"^/api/v1/events/([^/]+)/fail$")
_VOID = re.compile(r"^/api/v1/events/([^/]+)/void$")
_GET_BUDGET = re.compile(r"^/api/v1/budgets/([^/]+)$")
_TREE = re.compile(r"^/api/v1/budgets/([^/]+)/tree$")


def _f(x):
    return float(x) if x is not None else None


def _feature_for(path: str) -> str:
    if "delegation" in path:
        return "Delegation tokens"
    if "subscription" in path or "entitlement" in path:
        return "Subscriptions & entitlements"
    if "webhook" in path:
        return "Webhooks"
    if "replay" in path:
        return "Spend replay"
    return f"This operation ({path})"


def _snap_json(snap: dict) -> dict:
    return {
        "totalLimit": _f(snap["total_limit"]),
        "quantitySpent": _f(snap["quantity_spent"]),
        "quantityReserved": _f(snap["quantity_reserved"]),
        "availableQuantity": _f(snap["available"]),
        "status": snap["status"],
    }


class EmbeddedBackend:
    backend = "embedded"

    def __init__(self, database: str = ":memory:"):
        self.database = database
        self.engine = LiteEngine(database)
        # session_token → budget_id is PERSISTED in SQLite (engine.put/resolve_session_token),
        # not held in memory — so budgets survive a restart and stay authorizable (multi-day budgets).

    # The transport contract: same signature surface as the HTTP _request path.
    def request(self, method: str, path: str, json: Optional[dict] = None,
                params=None, headers: Optional[dict] = None) -> dict:
        body = json or {}
        headers = headers or {}

        if method == "POST" and path == "/api/v1/budgets":
            return self._create_budget(body)
        if method == "POST" and path == "/api/v1/authorize":
            return self._authorize(body, headers.get("X-Session-Token"))
        if method == "POST":
            m = _CONFIRM.match(path)
            if m:
                return self._event_json(self.engine.confirm(
                    event_id=m.group(1), confirmed_quantity=body["confirmedQuantity"]))
            m = _FAIL.match(path)
            if m:
                return self._event_json(self.engine.fail(event_id=m.group(1), reason=body.get("reason")))
            m = _VOID.match(path)
            if m:
                return self._event_json(self.engine.void(event_id=m.group(1), reason=body.get("reason")))
        if method == "GET":
            m = _TREE.match(path)
            if m:
                return self._tree_json(self.engine.get_tree(m.group(1)))
            m = _GET_BUDGET.match(path)
            if m:
                return self._budget_json(m.group(1), {}, None, self.engine.get_budget(m.group(1)))

        # anything else is a fleet/server-only endpoint
        raise FiGuardCapabilityError(_feature_for(path))

    # -- handlers ----------------------------------------------------------------

    def _reject_server_only_create_opts(self, body: dict) -> None:
        """Refuse create-time options embedded can't enforce — so a budget never *looks*
        configured (category caps, shadow mode, anomaly) while silently enforcing nothing of
        the sort. Mirrors the runtime capability boundary. (soft_limit is advisory → allowed.)"""
        if body.get("allocations"):
            raise FiGuardCapabilityError("Category allocations")
        trust_mode = body.get("trustMode")
        if trust_mode and str(trust_mode).upper() != "STRICT":
            raise FiGuardCapabilityError("Shadow / trust modes")
        if body.get("anomalyDetectionEnabled"):
            raise FiGuardCapabilityError("Anomaly detection")

    def _create_budget(self, body: dict) -> dict:
        self._reject_server_only_create_opts(body)
        bid = self.engine.create_budget(
            total_limit=body["totalLimit"],
            user_id=body.get("userId"),
            currency=body.get("currency"),
            unit=body.get("unit") if body.get("unit") else (None if body.get("currency") else "usd"),
            expires_at=body.get("expiresAt"),   # enforced by the engine when present
            max_transaction_quantity=body.get("maxTransactionQuantity"),
            intent_tags=body.get("intentTags"),
            entity_dedup_enabled=bool(body.get("entityDedupEnabled", False)),
            velocity_max_per_minute=body.get("velocityMaxPerMinute"),
            velocity_max_amount_per_hour=body.get("velocityMaxAmountPerHour"),
            velocity_max_per_day=body.get("velocityMaxPerDay"),
        )
        token = "st_" + uuid.uuid4().hex
        self.engine.put_session_token(token, bid)
        return self._budget_json(bid, body, token, self.engine.get_budget(bid))

    def _authorize(self, body: dict, session_token: Optional[str]) -> dict:
        budget_id = self.engine.resolve_session_token(session_token)
        if budget_id is None:
            raise FiGuardApiError(401, "INVALID_SESSION_TOKEN")
        try:
            r = self.engine.authorize(
                budget_id=budget_id,
                amount=body["requestedQuantity"],
                idempotency_key=body.get("idempotencyKey"),
                entity_id=body.get("entityId"),
                reserve=body.get("reserve", True),
                currency=body.get("currency"),
                claimed_category=body.get("claimedCategory"),
                parent_event_id=body.get("parentEventId"),
                intent_context=body.get("intentContext"),
                agent_id=body.get("agentId"),
                action_type=body.get("actionType"),
                description=body.get("description"),
            )
        except InvalidParentError as e:
            raise FiGuardApiError(400, str(e))
        return {
            "eventId": r.get("event_id"),
            "decision": r["decision"],
            "approvedQuantity": _f(r.get("approved_quantity")),
            "denialReason": r.get("denial_reason"),
            "denialMessage": r.get("denial_message"),
            "originalEventId": r.get("original_event_id"),
            "budgetSnapshot": _snap_json(r["budget"]),
        }

    def _event_json(self, r: dict) -> dict:
        return {
            "id": r["event_id"],
            "decision": r["decision"],
            "requestedQuantity": _f(r.get("requested_quantity")),
            "confirmedQuantity": _f(r.get("confirmed_quantity")),
            "currency": r.get("currency"),
            "createdAt": r.get("created_at") or "",
        }

    # Reshape the engine's tree (snake_case event dicts, nested {event, children}) into the
    # server's GET /tree JSON (camelCase) so the SDK's _parse_tree_node consumes either backend
    # identically. Recurses depth-first; the engine guarantees a forest (no cycles).
    def _tree_json(self, tree: dict) -> dict:
        return {
            "roots": [self._tree_node_json(n) for n in tree["roots"]],
            "totalEvents": tree["total_events"],
        }

    def _tree_node_json(self, node: dict) -> dict:
        return {
            "event": self._tree_event_json(node["event"]),
            "children": [self._tree_node_json(c) for c in node["children"]],
        }

    def _tree_event_json(self, e: dict) -> dict:
        return {
            "id": e["id"],
            "decision": e["decision"],
            "requestedQuantity": _f(e.get("requested_quantity")),
            "confirmedQuantity": _f(e.get("confirmed_quantity")),
            "currency": e.get("currency"),
            "entityId": e.get("entity_id"),
            "claimedCategory": e.get("claimed_category"),
            "agentId": e.get("agent_id"),
            "actionType": e.get("action_type"),
            "description": e.get("description"),
            "denialReason": e.get("denial_reason"),
            "parentEventId": e.get("parent_event_id"),
            "createdAt": e.get("created_at") or "",
        }

    def _budget_json(self, bid: str, create_body: dict, token: Optional[str], snapshot: dict) -> dict:
        # Metadata is read from the snapshot (which now carries it) so a GET is identical to the
        # CREATE response; create_body only supplies userId (not persisted on the budget row).
        return {
            "id": bid,
            "userId": snapshot.get("user_id") or create_body.get("userId", "embedded"),
            "totalLimit": _f(snapshot["total_limit"]),
            "quantitySpent": _f(snapshot["quantity_spent"]),
            "quantityReserved": _f(snapshot["quantity_reserved"]),
            "availableQuantity": _f(snapshot["available"]),
            "status": snapshot["status"],
            "expiresAt": snapshot.get("expires_at") or "",
            "currency": snapshot.get("currency"),
            "unit": snapshot.get("unit") or (None if snapshot.get("currency") else "usd"),
            "intentTags": snapshot.get("intent_tags"),
            "maxTransactionQuantity": _f(snapshot.get("max_transaction_quantity")),
            "velocityMaxPerMinute": snapshot.get("velocity_max_per_minute"),
            "velocityMaxAmountPerHour": _f(snapshot.get("velocity_max_amount_per_hour")),
            "velocityMaxPerDay": snapshot.get("velocity_max_per_day"),
            "tokens": [{"sessionToken": token, "category": None}] if token else None,
        }

File: sdk/typescript/src/_embedded/backend.ts
/**
 * Embedded backend — serves the FiGuard REST contract in-process, no server.
 *
 * Implements the same (method, path, body, headers) → server-shaped JSON contract the HTTP
 * transport provides, so FiGuardClient's methods work UNCHANGED on either backend. Optional
 * JSON-file persistence (Node fs) so budgets survive restarts. Fleet endpoints throw
 * FiGuardCapabilityError. Mirrors the Python embedded backend.
 */

import { FiGuardApiError, FiGuardCapabilityError } from "../errors";
import { EventResult, EventRow, InvalidParentError, LiteEngine, Snapshot, TreeNode, TreeResult } from "./engine";

function genToken(): string {
  return "st_" + Array.from({ length: 32 }, () => ((Math.random() * 16) | 0).toString(16)).join("");
}

// Session tokens are stored by SHA-256 hash (never raw), mirroring the server — a leaked
// JSON store yields no live tokens.
function hashToken(token: string): string {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return require("crypto").createHash("sha256").update(token).digest("hex");
}

function featureFor(path: string): string {
  if (path.includes("delegation")) return "Delegation tokens";
  if (path.includes("subscription") || path.includes("entitlement")) return "Subscriptions & entitlements";
  if (path.includes("webhook")) return "Webhooks";
  if (path.includes("replay")) return "Spend replay";
  return `This operation (${path})`;
}

export class EmbeddedBackend {
  readonly backend = "embedded";
  private engine = new LiteEngine();
  private tokens = new Map<string, string>(); // sha256(sessionToken) -> budgetId; PERSISTED
  private dbPath?: string;

  constructor(database = ":memory:") {
    if (database && database !== ":memory:") {
      this.dbPath = database;
      this.loadFromDisk();
    }
  }

  // Same surface as the HTTP request path; returns a resolved Promise.
  async request(
    method: string,
    path: string,
    opts: { body?: Record<string, unknown>; headers?: Record<string, string> } = {},
  ): Promise<Record<string, unknown>> {
    const body = (opts.body || {}) as Record<string, unknown>;
    const headers = opts.headers || {};
    let g: RegExpMatchArray | null;

    if (method === "POST" && path === "/api/v1/budgets") return this.persisted(this.createBudget(body));
    if (method === "POST" && path === "/api/v1/authorize")
      return this.persisted(this.authorize(body, headers["X-Session-Token"]));
    if (method === "POST" && (g = path.match(/^\/api\/v1\/events\/([^/]+)\/confirm$/)))
      return this.persisted(this.eventJson(this.engine.confirm(g[1], Number(body["confirmedQuantity"]))));
    if (method === "POST" && (g = path.match(/^\/api\/v1\/events\/([^/]+)\/fail$/)))
      return this.persisted(this.eventJson(this.engine.fail(g[1])));
    if (method === "POST" && (g = path.match(/^\/api\/v1\/events\/([^/]+)\/void$/)))
      return this.persisted(this.eventJson(this.engine.void(g[1])));
    if (method === "GET" && (g = path.match(/^\/api\/v1\/budgets\/([^/]+)\/tree$/)))
      return this.treeJson(this.engine.getTree(g[1]));
    if (method === "GET" && (g = path.match(/^\/api\/v1\/budgets\/([^/]+)$/)))
      return this.budgetJson(g[1], {}, undefined, this.engine.getSnapshot(g[1]));

    throw new FiGuardCapabilityError(featureFor(path));
  }

  // Refuse create-time options embedded can't enforce — so a budget never *looks* configured
  // (category caps, shadow mode, anomaly) while silently enforcing nothing. Mirrors the runtime
  // capability boundary. (softLimit is advisory → allowed.)
  private rejectServerOnlyCreateOpts(body: Record<string, unknown>): void {
    const allocations = body["allocations"] as unknown[] | undefined;
    if (allocations && allocations.length) throw new FiGuardCapabilityError("Category allocations");
    const trustMode = body["trustMode"] as string | undefined;
    if (trustMode && trustMode.toUpperCase() !== "STRICT") throw new FiGuardCapabilityError("Shadow / trust modes");
    if (body["anomalyDetectionEnabled"]) throw new FiGuardCapabilityError("Anomaly detection");
  }

  private createBudget(body: Record<string, unknown>): Record<string, unknown> {
    this.rejectServerOnlyCreateOpts(body);
    const id = this.engine.createBudget({
      totalLimit: body["totalLimit"] as number,
      userId: body["userId"] as string | undefined,
      currency: body["currency"] as string | undefined,
      unit: (body["unit"] as string) ?? (body["currency"] ? undefined : "usd"),
      expiresAt: (body["expiresAt"] as string) ?? null,
      maxTransactionQuantity: (body["maxTransactionQuantity"] as number) ?? null,
      intentTags: (body["intentTags"] as string[]) ?? null,
      entityDedupEnabled: !!body["entityDedupEnabled"],
      velocityMaxPerMinute: (body["velocityMaxPerMinute"] as number) ?? null,
      velocityMaxAmountPerHour: (body["velocityMaxAmountPerHour"] as number) ?? null,
      velocityMaxPerDay: (body["velocityMaxPerDay"] as number) ?? null,
    });
    const token = genToken();
    this.tokens.set(hashToken(token), id);
    return this.budgetJson(id, body, token, this.engine.getSnapshot(id));
  }

  private authorize(body: Record<string, unknown>, sessionToken?: string): Record<string, unknown> {
    const budgetId = sessionToken ? this.tokens.get(hashToken(sessionToken)) : undefined;
    if (!budgetId) throw new FiGuardApiError(401, "INVALID_SESSION_TOKEN");
    let r;
    try {
      r = this.engine.authorize({
        budgetId,
        amount: body["requestedQuantity"] as number,
        idempotencyKey: body["idempotencyKey"] as string | undefined,
        entityId: body["entityId"] as string | undefined,
        reserve: body["reserve"] !== false,
        currency: body["currency"] as string | undefined,
        claimedCategory: body["claimedCategory"] as string | undefined,
        parentEventId: body["parentEventId"] as string | undefined,
        intentContext: body["intentContext"] as string | undefined,
        agentId: body["agentId"] as string | undefined,
        actionType: body["actionType"] as string | undefined,
        description: body["description"] as string | undefined,
      });
    } catch (e) {
      if (e instanceof InvalidParentError) throw new FiGuardApiError(400, String(e.message));
      throw e;
    }
    return {
      eventId: r.eventId ?? null,
      decision: r.decision,
      approvedQuantity: r.approvedQuantity ?? null,
      denialReason: r.denialReason ?? null,
      denialMessage: r.denialMessage ?? null,
      originalEventId: r.originalEventId ?? null,
      budgetSnapshot: r.snapshot,
    };
  }

  private eventJson(r: EventResult): Record<string, unknown> {
    return {
      id: r.eventId,
      decision: r.decision,
      requestedQuantity: r.requestedQuantity,
      confirmedQuantity: r.confirmedQuantity ?? null,
      currency: r.currency ?? null,
      createdAt: r.createdAt || "",
    };
  }

  // Reshape the engine's forest into the server's GET /tree JSON so makeSpendTreeNode consumes
  // either backend identically. Recurses depth-first; the engine guarantees a forest (no cycles).
  private treeJson(tree: TreeResult): Record<string, unknown> {
    return {
      roots: tree.roots.map((node) => this.treeNodeJson(node)),
      totalEvents: tree.totalEvents,
    };
  }
  private treeNodeJson(node: TreeNode): Record<string, unknown> {
    return {
      event: this.treeEventJson(node.event),
      children: node.children.map((c) => this.treeNodeJson(c)),
    };
  }
  private treeEventJson(e: EventRow): Record<string, unknown> {
    return {
      id: e.id,
      decision: e.decision,
      requestedQuantity: e.requested / 10_000,
      confirmedQuantity: e.confirmed != null ? e.confirmed / 10_000 : null,
      currency: e.currency ?? null,
      entityId: e.entityId ?? null,
      claimedCategory: e.claimedCategory ?? null,
      agentId: e.agentId ?? null,
      actionType: e.actionType ?? null,
      description: e.description ?? null,
      denialReason: e.denialReason ?? null,
      parentEventId: e.parentEventId ?? null,
      createdAt: e.createdAt || "",
    };
  }

  private budgetJson(
    id: string, createBody: Record<string, unknown>, token: string | undefined, snap: Snapshot,
  ): Record<string, unknown> {
    // Metadata is read from the snapshot (which now carries it) so a GET is identical to the
    // CREATE response; createBody only supplies userId (not persisted on the budget row).
    return {
      id,
      userId: snap.userId ?? (createBody["userId"] as string) ?? "embedded",
      totalLimit: snap.totalLimit,
      quantitySpent: snap.quantitySpent,
      quantityReserved: snap.quantityReserved,
      availableQuantity: snap.availableQuantity,
      status: snap.status,
      expiresAt: snap.expiresAt ?? "",
      currency: snap.currency ?? null,
      unit: snap.unit ?? (snap.currency ? null : "usd"),
      intentTags: snap.intentTags ?? null,
      maxTransactionQuantity: snap.maxTransactionQuantity ?? null,
      velocityMaxPerMinute: snap.velocityMaxPerMinute ?? null,
      velocityMaxAmountPerHour: snap.velocityMaxAmountPerHour ?? null,
      velocityMaxPerDay: snap.velocityMaxPerDay ?? null,
      tokens: token ? [{ sessionToken: token, category: null }] : null,
    };
  }

  // -- optional JSON-file persistence (Node) --------------------------------

  private persisted<T extends Record<string, unknown>>(v: T): T {
    if (this.dbPath) this.saveToDisk();
    return v;
  }
  private loadFromDisk(): void {
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const fs = require("fs");
      if (!fs.existsSync(this.dbPath)) return;
      const blob = JSON.parse(fs.readFileSync(this.dbPath, "utf8"));
      // Persist BOTH the engine state and the token map, so a restarted process can resolve
      // a budget's session token (without this, persisted budgets become unauthorizable).
      this.engine.load(blob.engine ?? blob);            // blob.engine (new) | bare state (legacy)
      this.tokens = new Map(Object.entries(blob.tokens ?? {}));
    } catch { /* fresh store */ }
  }
  private saveToDisk(): void {
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const fs = require("fs");
      const path = require("path");
      fs.mkdirSync(path.dirname(this.dbPath), { recursive: true });
      fs.writeFileSync(this.dbPath, JSON.stringify({
        engine: this.engine.dump(),
        tokens: Object.fromEntries(this.tokens),
      }));
    } catch { /* non-fatal */ }
  }
}
